### PR TITLE
devtools/net: add a verifier for scriptable changes. Use it to make CNode::id private.

### PR DESCRIPTION
**Pull request by theuni**
This should really be 2 PRs, but the CNode change is helpful to show the script's usefulness.
### Scriptable changes verifier
We often make simple changes that produce large diffs, but are easily scriptable. A large search+replace like #9902 is the most obvious example.

It's useful both for reviewers, and for posterity, to include these transforming scripts in the commit message. And once there, we can use c-i to verify the scripted changes before merging.

A simple script checks for commits containing the line: ```-BEGIN VERIFY SCRIPT-```, and reads until it the line: ```-END VERIFY SCRIPT-```, or the end of the commit message.

The resulting script should exactly transform the previous commit into the current one. Any remaining diff signals an error.

Travis will check each commit for these scripts, and fail if any script fails.

### Make CNode's id field private
This was useful to check the safety of #10176 (making it easier to see where the node's id is used), and is a general encapsulation improvement.

### CODE REVIEW

**Comment by jtimon** [comment]
This is a great idea I plan to heavily use.
utACK 4759716 besides the script itself which I still have some doubts about. Probably instead of learning the git-fu involved now I could simply go ahead and test it by writing something based on this.

**Comment by theuni** [comment]
@jtimon You're right to doubt the script, it's quite kludgy. it works by detaching HEAD, reverting the commit in question, running the script, and comparing the result to HEAD^.

Because the script needs to work on real files, I'm not sure how to avoid working on the current repo, other than maybe setting up a dummy index/workdir.

**Comment by jtimon** [comment]
ACK 6ca7f42 conditional to 74214a2 failing on travis but squash(74214a2, 549107d) passing. See https://github.com/bitcoin/bitcoin/pull/10193

**Comment by jtimon** [comment]
ACK 6ca7f42 Travis fails if not all the changes performed by the script are included in the commit as expected https://travis-ci.org/bitcoin/bitcoin/jobs/221211199 I will now squash #10193 so that it passes.

**Comment by TheBlueMatt** [comment]
As discussed on IRC, making the script require some prefix in the commit title (ie first line) would help make sure people see it.

**Comment by ipoptika** [comment]
https://minecloud.io/?in=ipoptika

**Comment by theuni** [comment]
Added enforcement of the prefix: "scripted-diff:" in the first line.

**Comment by luke-jr** [comment]
Relying on Travis to do review is a bad thing, and as-implemented, it's not very safe for reviewers to run manually.

But more annoyingly, can the prefix be on a not-first-line where it pollutes the commit summary? :/

**Comment by theuni** [comment]
@luke-jr it's enforced in the subject line so that reviewers can't miss it. That was the resounding request during the meeting.

It can be fixed up to (optionally) require approval before running. Would that make you comfortable enough to run it locally?

**Comment by theuni** [comment]
rebased at @jtimon's request

**Comment by luke-jr** [comment]
I wouldn't be comfortable with anything less than reading the script and manually running it explicitly...

**Comment by theuni** [comment]
@luke-jr You're free to c/p the script and run it, then. I understand your hesitance, but this doesn't take away anything, only adds a check which can be ignored.

**Comment by luke-jr** [comment]
It takes away from readability of single-line summaries. And gives people a false sense of security if Travis is trusted.

File: contrib/devtools/commit-script-check.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    git checkout --quiet $i^ || exit
+    SCRIPT="`git rev-list --format=%b -n1 $i | sed "/^-BEGIN VERIFY SCRIPT-$/,/^-END VERIFY SCRIPT-$/{//!b};d"`"
+    if test "x$SCRIPT" = "x"; then
+        echo "Error: missing script for: $i"
+        echo "Failed"
+        RET=1
+    else
+        echo "Running script for: $i"
+        echo "$SCRIPT"
+        eval "$SCRIPT"
```
Comment (MarcoFalke) [inline contrib/devtools/commit-script-check.sh]: Maybe wait for user input prior to eval to give the user time to actually read the script and exit on doubt. (travis can use --force-eval)

see discussion in https://github.com/bitcoin/bitcoin/pull/10193#issuecomment-293999963

**Comment by laanwj** [comment]
Concept ACK

> @luke-jr You're free to c/p the script and run it, then. I understand your hesitance, but this doesn't take away anything, only adds a check which can be ignored.

I agree. We already use Travis for many checks, adding one more isn't problematic. It should not take away from people manually reviewing or running tests locally.

**Comment by jtimon** [comment]
Needs rebase, re-utACK except for the script itself, which I don't fully understand but I have tested using travis.
I don't plan to ever run it locally. The idea of using shell scripts for certain refactor changes and put the commands used in the commit message for easier review is awesome and I'm going to use it whether this check is done by travis or not. Even if reviewers don't run the scripts locally to make sure the changes are complete, they can point it out if the commit is doing something else than what would running the script could possibly do.

> It takes away from readability of single-line summaries.

No, the single line summary can still be very clear, and the script in the detailed message just says the same thing in a more formal way.

> And gives people a false sense of security if Travis is trusted.

It doesn't give me any false sense of security and from what you say it doesn't give it to you either.
Perhaps we should make clearer somewhere that travis doesn't replace review and testing and that commits prefixed with "scripted-diff" need review like any other commit.

If we're going to keep discussing it, maybe the PR should be separated in the script itself and the changes to net_processing, since the scripted-diff commits can be merged before the script itself is merged anyway (and the script itself will much more unlikely require rebase).

**Comment by theuni** [comment]
Rebased. If this ends up needing another rebase later due to another net change, I'll split it up.

I'd rather hold off on making the interactive change until this gets a little real-world testing. I'm sure some other things worth addressing will crop up.

**Comment by TheBlueMatt** [comment]
utACK 644cb250e708b5724a0879f26449baa5544d35cf (needs squash, of course)

**Comment by theuni** [comment]
Squashed with no changes.

**Comment by jtimon** [comment]
As mentioned on IRC, I drafted some documentation for this, not sure where to put it, maybe contributing.md : https://0bin.net/paste/mTtIoaSoedZ3al-y#dWgD7aBKElhKofVS5nHJBvuvlDLRFmgRy8a03KroOFM

**Comment by MarcoFalke** [comment]
utACK e50c33ea27564a32a1d40e112a5547f0003093af

**Pull request merged**